BJS2-26818 Организация события

Murzin34 opened the pull request:
Ссылка на Jira: https://faang-school.atlassian.net/browse/BJS2-26818
При маппинге в контроллере не отрабатывают корректно  методы `createEvent()` и `updateEvent()` из-за полей `Long ownerId` в `EventDto` и `User owner` в `Event`. Пока так и не догадался как это исправить(возможно надо задействовать UserContext), поэтому маппинг пока осуществляю в сервисе.

Murzin34 committed:
BJS2-26818 Организация события

Murzin34 committed:
Фикс.
Изменил реализацию фильтров.

Murzin34 committed:
Фикс.
Небольшой фикс EventFilterDto, EventTypeFilter, EventValidator.java. Поменял реализацию фильтров EventRelatedAllSkillsFilter и EventRelatedAnySkillsFilter на более простую.

Murzin34 committed:
Тесты.

Murzin34 committed:
Добавил кастомные маперы.
EventCustomMapper и SkillCustomMapper, но пока их не задействую.

Murzin34 committed:
Фикс.
Фикс функционала и тестов

Alginin commented on src/main/java/school/faang/user_service/controller/event/EventController.java:
```diff
@@ -0,0 +1,64 @@
+        return eventService.getEvent(eventId);
+    }
+
+    @PostMapping("/filters")
```
Может быть  должен использоваться @GetMapping ?

Murzin34 replied:
По опыту предыдущего буткемпа помню, что при использование фильтров нужно ставить `@PostMapping`.

20cognosce replied:
Get по смыслу конечно лучше, но  у него не должно быть тела по спецификации
Когда много фильтров, то их передаем в теле, так что уж да, нужен Post

Peregruzochka commented on src/main/java/school/faang/user_service/validation/event/EventValidator.java:
```diff
@@ -0,0 +1,45 @@
+
+    public void relatedSkillsValidation(EventDto eventDto) {
+        List<SkillDto> ownerSkills = skillService.getUserSkillsList(eventDto.getOwnerId());
+        if (!new HashSet<>(ownerSkills).containsAll(eventDto.getRelatedSkills())) {
```
Можно создать HashSet ранее, для повышения читаемости

Murzin34 replied:
Поправил.

Peregruzochka commented on src/test/java/school/faang/user_service/service/event/EventServiceTest.java:
```diff
@@ -0,0 +1,213 @@
+            assertNotNull(result);
+            assertEquals(eventDto, result);
+
+            verify(eventValidator, atLeastOnce()).eventDatesValidation(eventDto);
+            verify(eventValidator, atLeastOnce()).relatedSkillsValidation(eventDto);
+            verify(userService, atLeastOnce()).findUserById(user.getId());
+            verify(eventRepository, atLeastOnce()).save(event);
```
Как будто нет смысла вызвать verify для всех методов. Можно оставить verify для save, т.к. он он отработает только в том случае, если все предыдущие методы будут вызваны.

Murzin34 replied:
Надо спросить, но в целом ты прав. Если в указанные в verify методы не отработают тест упадет раньше.

Peregruzochka commented on src/main/java/school/faang/user_service/dto/event/EventDto.java:
```diff
@@ -0,0 +1,60 @@
+    private Long ownerId;
+
+    @NotNull(message = "Event related_skills cannot be null.")
+    private List<SkillDto> relatedSkills;
```
Я бы использовал List<Long> relatedSkillIds, т.к. мы можем передавать сюда только те скиллы, которые есть в базе данных. (Ты как раз проверяешь существуют ли эти скиллы при создании нового евента)

Murzin34 replied:
Передаю всю сущность SkillDto т.к. в фильтрах использую проверку по имени скилла, а не id. Хотя в целом стоит пожалуй сделать как ты говоришь когда добью маппинг в контроллере.

Peregruzochka commented on src/test/java/school/faang/user_service/service/event/EventServiceTest.java:
```diff
@@ -0,0 +1,213 @@
+        }
+
+        @Test
+        public void testDeleteEvent_Success() {
+            eventService.deleteEvent(eventDto.getId());
+
+            verify(eventRepository, atLeastOnce()).deleteById(eventDto.getId());
+        }
```
Я бы добавил when(findEventById(eventId)).thenReturn(true), т.к. необходимо, чтобы внутри delete метода валидация прошла успешно

Murzin34 replied:
Косяк, валидацию добавил в последний момент и про тест забыл. Поправил.

Peregruzochka commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    @Transactional
+    public EventDto updateEvent(EventDto eventDto) {
+        eventValidator.eventDatesValidation(eventDto);
+        eventValidator.relatedSkillsValidation(eventDto);
+        eventValidator.eventExistByDtoValidation(eventDto);
+        return saveEvent(eventDto);
+    }
```
Тут не реализована логика, которую нужно было сделать по заданию:
![image](https://github.com/user-attachments/assets/fc5c8b54-38bf-41f8-ad5e-b7285083d519)

Murzin34 replied:
Конкретно эта валидация реализована в `relatedSkillsValidation()`, но я помню, что мы с тобой говорили о какой-то упущенной мной валидации.

Murzin34 replied:
Я помню мы с тобой говорили про упущенную мной валидацию, но конкретно здесь валидация реализована в `relatedSkillsValidation() `,а в п.3 таски валидация происходит на уровне ДТО.

Peregruzochka commented on src/main/java/school/faang/user_service/validation/event/EventValidator.java:
```diff
@@ -0,0 +1,45 @@
+        }
+    }
+
+    public void eventExistByDtoValidation(EventDto eventDto) {
+        Long id = eventDto.getId();
+        if (!eventRepository.existsById(id)) {
+            throw new DataValidationException("Event ID: " + id + " dont exist");
+        }
+    }
+    public void eventExistByIdValidation(Long id) {
+        if (!eventRepository.existsById(id)) {
+            throw new DataValidationException("Event ID: " + id + " dont exist");
+        }
+    }
```
Произошло дублирование кода. Возможно стоит подумать как использовать этот метод в разных ситуациях.

Murzin34 replied:
Поправил. Оставил `eventExistByIdValidation()`

Peregruzochka commented on src/main/java/school/faang/user_service/controller/event/EventController.java:
```diff
@@ -0,0 +1,64 @@
+        return eventService.updateEvent(eventDto);
+    }
+
+    @GetMapping("/owner/{userId}")
+    public List<EventDto> getEventsOwner(@PathVariable @Positive long userId) {
+        return eventService.getEventsOwner(userId);
+    }
+
+    @GetMapping("/participant/{userId}")
+    public List<EventDto> getEventParticipants(@PathVariable @Positive long userId) {
+        return eventService.getEventParticipants(userId);
+    }
```
Имеет смысл использовать не `@PathVariable`, а `@RequestParam`, который будет задается в Http-запросе.
Таким образом, нет необходимости прописывать urls в данных мапперах.
Пример того как это делал я в Postman:
![image](https://github.com/user-attachments/assets/a476c40f-6a55-40e8-af05-3d2dc0a186e5)

Скорее всего валидация `@Positive` избыточна. (этот код будут использовать другие разработчики, а не users)
@mrkorolev

Murzin34 replied:
Заменил на @RequestParam. По поводу избыточной валидации надо спросить на созвоне.

Alginin commented on src/test/java/school/faang/user_service/validation/event/EventValidatorTest.java:
```diff
@@ -0,0 +1,130 @@
+        }
+
+        @Test
+        public void testEventExistByIdValidation_NotFound_throwDataValidationException() {
```
В лекциях говорилось что использовать snake case в названиях методов не лучшая практика, а так же перемешивание camelCase + snake_case.

Murzin34 replied:
Именование тестов вопрос холиварный) надо на созвоне спросить и тогда переименую.

20cognosce replied:
Это не snake_case, это 3 стадии, описывающие тест
что делаем
что получаем
как реагируем

Peregruzochka reviewed (commented):
Мне не совсем понятно, является ли это хорошей практикой, когда мы для работы с конкретным классом репозитория, работаем только с соответствующим сервисом:
Например: `UserRepository` <-> `UserService`
Могу ли я инжектить в класс UserService другие Repository (например GoalRepository, SkillRepository)?

mrkorolev commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,101 @@
+    private final EventRepository eventRepository;
+    private final EventCustomMapper eventMapper;
+    private final UserService userService;
+    private final EventValidator eventValidator;
```
Как по мне, избыточно выносить логику валидации за рамки сервиса по нескольким причинам:
- валидатор использует те же зависимости, что и сервис (одни и те же зависимости разнесены по двум разным бинам)
- логика сервиса разнесена по разным местам (а валидации - бизнес логика твоего сервиса)
- это не удобно читать и тестировать (постоянно перепрыгивать из одного файла в другой)

Murzin34 replied:
Да, по сути валидация в отдельном классе идет как дополнение к сервису, но вопрос все равно спорный, надо будет спросить на созвоне.

Murzin34 committed:
Фикс после кросс-ревью.

Murzin34 committed:
Повесил TODO

Murzin34 committed:
Заменил @Data на @Getter и @Setter в DTO

20cognosce commented on src/main/java/school/faang/user_service/controller/event/EventController.java:
```diff
@@ -0,0 +1,66 @@
+        return eventService.updateEvent(eventDto);
+    }
+
+    @GetMapping("/owner")
+    public List<EventDto> getEventsOwner(@RequestParam @Positive long ownerId) {
+        return eventService.getEventsOwner(ownerId);
+    }
+
+    @GetMapping("/participant")
+    public List<EventDto> getEventParticipants(@RequestParam @Positive long participantsId) {
+        return eventService.getEventParticipants(participantsId);
+    }
```
Почему аннотации @Positive в методе? Валидация должна быть не здесь

Murzin34 replied:
Убрал `@positive` в контролере. Вот хотели спросить на созвоне про валидацию и в контролере и в ДТО да забыли) ну вообщем теперь понятно что это излишне.

20cognosce commented on src/main/java/school/faang/user_service/mapper/event/EventCustomMapper.java:
```diff
@@ -0,0 +1,72 @@
+
+@Component
+@RequiredArgsConstructor
+public class EventCustomMapper {
```
Норм, я на работе так же всё руками делаю

Murzin34 replied:
Спасибо) но я сделал только потому что ошибочно подумал что проблема мапинга в контроллерах именно в MapStruct.

20cognosce commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,23 @@
+    private final SkillRepository skillRepository;
+    private final SkillCustomMapper skillMapper;
+
+    public List<SkillDto> getUserSkillDtoList(Long userId) {
```
Для чего существует этот метод? Почему такое название

Murzin34 replied:
Метод использую для валидации методов EventService `createEvent()` и `updateEvent()`. 
В таске это пункты 1.6 и 5.5:  "Здесь нужно проверить, что пользователь, который создает событие, имеет навыки, связанные с событием." 
Называние Идея подсказала через ctrl+пробел "SkillDtoList" я дополнил дописав "getUser".

20cognosce replied:
То что сервис отдает DTO для другого сервиса может создать кучу неудобств

20cognosce replied:
Мне не нравится нейминг
Почему в нейминге ты пишешь и тип возвращаемых данных и коллекцию, какой смысл? Это итак есть в сигнатуре

Murzin34 replied:
Переименовал в getUserSkills и положил в EventService.

20cognosce commented on src/main/java/school/faang/user_service/service/UserService.java:
```diff
@@ -0,0 +1,20 @@
+
+@Service
+@RequiredArgsConstructor
+public class UserService {
+    private final UserRepository userRepository;
+
+    //TODO Murzin34 Задействовать обработчик ошибок
+    public User getUserById(Long userId) {
+        return userRepository.findById(userId).orElseThrow(() ->
+                new DataValidationException("User with ID: " + userId + " not found")
+        );
+    }
```
Я не понял, почему ты не используешь готовые сервисы других разработчиков, а пишешь свои? Твоя задача по ивентам

Murzin34 replied:
На данный момент в мастер-ветке нет функционала других ребят. По опыту предыдущего потока, нас, наоборот, пытались столкнуть лбами при мердже файлов, чтобы мы, взаимодействуя друг с другом, решали эту проблему.
Попробовал реализовать интерфейсы, как ты сказал, для фиксирования методов в "сторонних" сервисах, которые использую в своем "профильном" сервисе. Надеюсь, такая реализация устроит. Если нет, то поправлю.

20cognosce commented on src/main/java/school/faang/user_service/service/event/filters/EventStartDateAfterFilter.java:
```diff
@@ -0,0 +1,19 @@
+import school.faang.user_service.entity.event.Event;
+
+@Component
+public class EventStartDateAfterFilter implements EventFilter {
```
Что за жесть с фильтрами, сколько их? Зачем?

Murzin34 replied:
Нуу на каждое поле в ДТО по фильтру или по два) кроме owner, потому что под него отдельный контролер пишем (п.6 в таске). Это же все равно мое будущее CV, мое будущее резюме, хотя да, ты говорил что фильтры надо писать через нативный SQL, я все равно доволен собой что разобрался в вопросе)
В общем то тут 4 ВИДА фильтров: 
1. Фильтры на поля сущности String, их 3 шт.
2. Фильтры на поля сущности LocalDateTime их 4 шт.
3. Фильтры на поля сущности int, их 2 шт.
4. Фильтры на поля сущности enum, их 4 шт.

Функционал и тесты для одних и тех же ВИДОВ одинаковый, главное не напутать с вызовами методов.

20cognosce commented on src/main/java/school/faang/user_service/validation/event/EventValidator.java:
```diff
@@ -0,0 +1,39 @@
+
+@Component
+@RequiredArgsConstructor
+public class EventValidator {
```
Он неотделимая часть сервиса, рядом в структуре их положи что ли

Murzin34 replied:
EventValidator положил в "service.event", пакедж "validation.event" удалил.

20cognosce commented on src/test/java/school/faang/user_service/controller/event/EventControllerTest.java:
```diff
@@ -0,0 +1,124 @@
+        assertNotNull(result);
+        assertEquals(eventDto, result);
+
+        verify(eventService, atLeastOnce()).createEvent(eventDto);
```
почему atLeastOnce
а если будет 2?

Murzin34 replied:
Вроде часто используют в VerificationMode=times(1), хотя вроде можно вообще обойтись без него. Писать на манер `verify(eventService).createEvent(eventDto);`, но я на всякий случай пишу atLeastOnce(). 
В некоторых случаях пишу times(*что-то*.size()) как например в EventServiceTest в тесте фильтров.

20cognosce commented on src/test/java/school/faang/user_service/mapper/event/EventCustomMapperTest.java:
```diff
@@ -0,0 +1,97 @@
+        eventDto = eventCustomMapper.toDto(event);
+
+        assertNotNull(eventDto);
+        assertEquals(event.getId(), eventDto.getId());
+        assertEquals(event.getTitle(), eventDto.getTitle());
+        assertEquals(event.getDescription(), eventDto.getDescription());
+        assertEquals(event.getStartDate(), eventDto.getStartDate());
+        assertEquals(event.getEndDate(), eventDto.getEndDate());
+        assertEquals(event.getLocation(), eventDto.getLocation());
+        assertEquals(event.getMaxAttendees(), eventDto.getMaxAttendees());
+        assertEquals(event.getOwner().getId(), eventDto.getOwnerId());
+        assertEquals(event.getType(), eventDto.getType());
+        assertEquals(event.getStatus(), eventDto.getStatus());
```
assertThat().recursiveComparison()

Murzin34 replied:
Спасибо! Поправил тесты.

20cognosce commented on src/test/java/school/faang/user_service/service/event/filters/EventMaxAttendeesLargerFilterTest.java:
```diff
@@ -0,0 +1,79 @@
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class EventMaxAttendeesLargerFilterTest {
```
Ты для каждого фильтра тест писал - это безумие

Murzin34 replied:
В общем то тут 4 ВИДА фильтров: 
1. Фильтры на поля сущности String, их 3 шт.
2. Фильтры на поля сущности LocalDateTime их 4 шт.
3. Фильтры на поля сущности int, их 2 шт.
4. Фильтры на поля сущности enum, их 4 шт.

Функционал и тесты для одних и тех же ВИДОВ одинаковый, главное не напутать с вызовами методов.

20cognosce commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,106 @@
+    private EventDto saveEvent(EventDto eventDto) {
+        Event event = eventMapper.toEntity(eventDto);
+
+        User newEventOwner = userService.getUserById(eventDto.getOwnerId());
+        event.setOwner(newEventOwner);
```
Вот ты юзаешь чужой сервис, реализовав его по-своему
Я вмерджу - а у другого чела конфликт будет

Murzin34 replied:
На данный момент в мастер-ветке нет функционала других ребят. По опыту предыдущего потока, нас, наоборот, пытались столкнуть лбами при мердже файлов, чтобы мы, взаимодействуя друг с другом, решали эту проблему.
Попробовал реализовать интерфейсы, как ты сказал, для фиксирования методов в "сторонних" сервисах, которые использую в своем "профильном" сервисе. Надеюсь, такая реализация устроит. Если нет, то поправлю.

20cognosce reviewed (changes requested):
Скажу больную для тебя вещь - но миллион фильтров и тележка тестов к ним в придачу были напрасны

Использование чужих сервисов убирай. Либо свяжись с ними и зафиксируй общие интерфейсы, а они потом свою реализацию сделают.

Ну и фильтры... 
Если у тебя душа за них болит - ладно, пусть будут

Murzin34 committed:
Фикс после ревью.
Удалил излишнюю валидацию в контролере. В EventDto и SkillDto добавил анотации @@ToString и @EqualsAndHashCode на отдельное поле т.к. нужны для тестов. Перенес валидатор в пакедж сервиса. Написал 2 интерфейса для пояснения какие методы использую в сторонних сервисах. Поправил тесты.

20cognosce commented on src/main/java/school/faang/user_service/service/event/EventService.java:
```diff
@@ -0,0 +1,105 @@
+public class EventService {
+    private final EventRepository eventRepository;
+    private final EventCustomMapper eventMapper;
+    private final UserService userService;
```
Используй UserRepository
UserService полностью удали

Murzin34 replied:
Удалил

20cognosce commented on src/main/java/school/faang/user_service/service/event/EventValidator.java:
```diff
@@ -0,0 +1,39 @@
+@RequiredArgsConstructor
+public class EventValidator {
+    private final EventRepository eventRepository;
+    private final SkillService skillService;
```
Используй SkillRepository
SkillService полностью удали

Murzin34 replied:
Удалил

20cognosce commented on src/main/java/school/faang/user_service/service/event/EventValidator.java:
```diff
@@ -0,0 +1,39 @@
+
+@Component
+@RequiredArgsConstructor
+public class EventValidator {
```
Назови его EventServiceHelper

Murzin34 replied:
Переименовал.

20cognosce commented on src/main/java/school/faang/user_service/service/event/filters/EventDescriptionFilter.java:
```diff
@@ -0,0 +1,21 @@
+import school.faang.user_service.entity.event.Event;
+
+@Component
+public class EventDescriptionFilter implements EventFilter {
```
Фильтры убрать из сервиса, вынести их к репозиториям

Murzin34 replied:
Убрал.

20cognosce commented on src/main/java/school/faang/user_service/mapper/SkillCustomMapper.java:
```diff
@@ -0,0 +1,22 @@
+import school.faang.user_service.entity.Skill;
+
+@Component
+public class SkillCustomMapper {
```
Почему SkillMapper в package school.faang.user_service.mapper;
а EventMapper в package school.faang.user_service.mapper.event?

Custom тоже лишнее слово

Murzin34 replied:
Сущность Skill лежит в school.faang.user_service.entity, а сущность Event school.faang.user_service.entity.event; 
С репозиториями также. Custom в имени убрал.

Murzin34 committed:
Фикс согласно ревью.
Удалил SkillService и UserService, их функционал перенес в EventService. Переименовал маперы и валидатор. Перенес фильтры в пакедж с репозиториями. Поправил тесты.
